source/server/connection_handler_impl.h

```diff
@@ -136,16 +138,34 @@ class ConnectionHandlerImpl : public Network::ConnectionHandler,
      */
     void newConnection(Network::ConnectionSocketPtr&& socket);
 
+    ActiveConnections& getOrCreateActiveConnections(const Network::FilterChain& filter_chain);
+
     ConnectionHandlerImpl& parent_;
     Network::ListenerPtr listener_;
     const std::chrono::milliseconds listener_filters_timeout_;
     const bool continue_on_listener_filters_timeout_;
     std::list<ActiveTcpSocketPtr> sockets_;
-    std::list<ActiveTcpConnectionPtr> connections_;
+    std::unordered_map<const Network::FilterChain*, ActiveConnectionsPtr> connections_by_context_;
 
     // The number of connections currently active on this listener. This is typically used for
     // connection balancing across per-handler listeners.
     std::atomic<uint64_t> num_listener_connections_{};
+    bool is_deleting_{false};
+  };
+
+  /**
+   * Wrapper for a group of active connections which are attached to the same filter chain context.
+   */
+  class ActiveConnections : public Event::DeferredDeletable {
+  public:
+    ActiveConnections(ActiveTcpListener& listener, const Network::FilterChain& filter_chain);
+    ~ActiveConnections();
+
+    // listener filterchain pair is the owner of the connections
+    ActiveTcpListener& listener_;
+    const Network::FilterChain& filter_chain_;
```

Comment:
Eventually this will need to have a shared_ptr to a filter chain, right?

Reply:
See the ownership of FilterChain comment: we don't need a shared_ptr here